components/body_sniffer/body_sniffer_url_loader.h

```diff
@@ -0,0 +1,136 @@
+      mojo::PendingReceiver<network::mojom::URLLoaderClient>
+          source_url_client_receiver,
+      mojo::ScopedDataPipeConsumerHandle body);
+
+ protected:
+  BodySnifferURLLoader(
+      base::WeakPtr<body_sniffer::BodySnifferThrottle> throttle,
+      const GURL& response_url,
+      mojo::PendingRemote<network::mojom::URLLoaderClient>
+          destination_url_loader_client,
+      scoped_refptr<base::SequencedTaskRunner> task_runner);
+
+  // network::mojom::URLLoaderClient implementation (called from the source of
+  // the response):
+  void OnReceiveEarlyHints(network::mojom::EarlyHintsPtr early_hints) override;
+  void OnReceiveResponse(network::mojom::URLResponseHeadPtr response_head,
+                         mojo::ScopedDataPipeConsumerHandle body) override;
+  void OnReceiveRedirect(
+      const net::RedirectInfo& redirect_info,
+      network::mojom::URLResponseHeadPtr response_head) override;
+  void OnUploadProgress(int64_t current_position,
+                        int64_t total_size,
+                        OnUploadProgressCallback ack_callback) override;
+  void OnReceiveCachedMetadata(mojo_base::BigBuffer data) override;
+  void OnTransferSizeUpdated(int32_t transfer_size_diff) override;
+  void OnStartLoadingResponseBody(
+      mojo::ScopedDataPipeConsumerHandle body) override;
+  void OnComplete(const network::URLLoaderCompletionStatus& status) override;
+
+  // network::mojom::URLLoader implementation (called from the destination of
+  // the response):
+  void FollowRedirect(
+      const std::vector<std::string>& removed_headers,
+      const net::HttpRequestHeaders& modified_headers,
+      const net::HttpRequestHeaders& modified_cors_exempt_headers,
+      const absl::optional<GURL>& new_url) override;
+  void SetPriority(net::RequestPriority priority,
+                   int32_t intra_priority_value) override;
+  void PauseReadingBodyFromNet() override;
+  void ResumeReadingBodyFromNet() override;
+
+  bool CheckBufferedBody(uint32_t readBufferSize);
+
+  virtual void OnBodyReadable(MojoResult) = 0;
+  virtual void OnBodyWritable(MojoResult) = 0;
+
+  void CompleteLoading(std::string body);
+  void CompleteSending();
+  virtual void CallClientComplete() = 0;
+  void SendReceivedBodyToClient();
+
+  void Abort();
+
+  base::WeakPtr<BodySnifferThrottle> throttle_;
+  GURL response_url_;
+
+  mojo::Receiver<network::mojom::URLLoaderClient> source_url_client_receiver_{
+      this};
+  mojo::Remote<network::mojom::URLLoader> source_url_loader_;
+  mojo::Remote<network::mojom::URLLoaderClient> destination_url_loader_client_;
+
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
+
+  enum class State { kWaitForBody, kLoading, kSending, kCompleted, kAborted };
+  State state_ = State::kWaitForBody;
+
+  absl::optional<network::URLLoaderCompletionStatus> complete_status_;
+
+  std::string buffered_body_;
+  size_t bytes_remaining_in_buffer_;
+
+  mojo::ScopedDataPipeConsumerHandle body_consumer_handle_;
```

Comment:
Do we really need all of these methods and field be protected, not private?
Some of them can easy moved to private.

Reply:
@atuchin-m I don't like exposing properties at all and would prefer we had protected methods for this stuff, but I think that can be pushed to a follow-up